src/cli/commands/lint.ts

```diff
@@ -0,0 +1,124 @@
+  public static description = 'lint a JSON/YAML document from a file or URL';
+
+  public static examples = [
+    `$ spectral lint .openapi.yaml
+linting ./openapi.yaml
+`,
+  ];
+
+  public static flags = {
+    help: flagHelpers.help({ char: 'h' }),
+    encoding: flagHelpers.string({
+      char: 'e',
+      default: 'utf8',
+      description: 'text encoding to use',
+    }),
+    format: flagHelpers.string({
+      char: 'f',
+      default: 'stylish',
+      description: 'formatter to use for outputting results',
+      options: ['json', 'stylish'],
+    }),
+    maxResults: flagHelpers.integer({
+      char: 'm',
+      description: '[default: all] maximum results to show',
+    }),
+    verbose: flagHelpers.boolean({
+      char: 'v',
+      description: 'increase verbosity',
+    }),
+  };
+
+  public static args = [{ name: 'source' }];
+
+  public async run() {
+    const { args, flags } = this.parse(Lint);
+
+    if (args.source) {
+      try {
+        await lint(args.source, flags, this);
+      } catch (ex) {
+        this.error(ex.message);
+      }
+    } else {
+      this.error('You must specify a document to lint');
+    }
+  }
+}
+
+async function lint(name: string, flags: any, command: Lint) {
+  command.log(`linting ${name}`);
+  let obj: any;
+  try {
+    obj = await readInputArguments(name, flags.encoding);
+  } catch (ex) {
+    throw new Error(`Could not parse ${name}: ${ex.message}`);
+  }
+
+  const spectral = new Spectral();
+  if (obj.data.swagger && obj.data.swagger === '2.0') {
+    command.log('OpenAPI 2.0 (Swagger) detected');
+    spectral.addFunctions(oas2Functions());
+    spectral.addRules(oas2Rules());
+  } else if (obj.data.openapi && typeof obj.data.openapi === 'string' && obj.data.openapi.startsWith('3.')) {
+    command.log('OpenAPI 3.x detected');
+    spectral.addFunctions(oas3Functions());
+    spectral.addRules(oas3Rules());
+  } else {
+    throw new Error('Input document specification type could not be determined');
+  }
+
+  let results = [];
+  try {
+    results = await spectral.run(obj.data);
+    if (results.length === 0) {
+      command.log('No errors or warnings found!');
+      return;
+    }
+  } catch (ex) {
+    process.exitCode = 2;
+    throw new Error(ex);
+  }
+  // Keep only the requested number of warnings
+  if (flags.maxResults) {
+    results = results.slice(0, flags.maxResults);
+  }
+  process.exitCode = 1;
+  const output = await formatOutput(results, flags.format);
+  command.log(output);
+}
+
+async function formatOutput(results: IRuleResult[], format: any) {
+  return {
+    json: () => json(results),
+    stylish: () => stylish(results),
+  }[format]();
+}
+
+async function readInputArguments(name: string, encoding: string) {
+  if (name.startsWith('http')) {
+    const result = await fetch(name);
+    return parseWithPointers(await result.text());
+  } else if (existsSync(name)) {
+    try {
+      return parseWithPointers(readFileSync(name, encoding));
```

Comment:
Doesn't hurt to be safe :).

Reply:
Well, true, don't negate that, but we don't handle the potential exception above, so if we decide to catch it here, should we catch the one above as well?

Reply:
It's not here because im worried about parseWithPointers, its here to catch readFileSync :)

Reply:
Ah, true! shouldn't use async version though? I see the method doesn't get called often, but still, the function is already async, so wouldn't hurt.

Reply:
```suggestion
      return parseWithPointers(await readFile(name, encoding));
```
That would be my proposal ^

Reply:
I'm not entirely sure if im just doing it wrong, but that doesnt seem to be how readFile works. 

```
fs.readFile('/etc/passwd', (err, data) => {
  if (err) throw err;
  console.log(data);
});
```

It does not return a promise, it asks for a callback to be passed in, so using await does not actually work.

I have seen folks wrapping it with promisify to solve this, but that seems strange because we dont need it to be async here... I am not sure there would be any benefit? Waiting for a promised thing to happen is syncronous, so would this not be the same?

Reply:
We use promisify everywhere. We will most likely switch to fs.promises eventually once stable.

Reply:
We use promisify everywhere. We will most likely switch to fs.promises eventually once stable.

Reply:
My two cents - I think `readFileSync` is fine for now. Can revisit later.

Reply:
Yeah, can do later.

Reply:
Ok, thanks!